packages/SwingSet/src/kernel/kernel.js

```diff
@@ -795,13 +795,17 @@ export default function buildKernel(kernelEndowments) {
     for (const vatID of kernelKeeper.getAllDynamicVatIDs()) {
       console.debug(`Loading dynamic vat ${vatID}`);
       const vatKeeper = kernelKeeper.allocateVatKeeperIfNeeded(vatID);
-      const {
-        source,
-        options: dynamicOptions,
-      } = vatKeeper.getSourceAndOptions();
-      // eslint-disable-next-line no-await-in-loop
-      await recreateVatDynamically(vatID, source, dynamicOptions);
-      // now the vatManager is attached and ready for transcript replay
+      if (vatKeeper.isDead()) {
+        kernelKeeper.forgetVat(vatID);
```

Comment:
Do we need to create a `{ dead: true }` record for `ephemeral.vats` here, to match the state of the world in the previous run (after the vat was terminated)? I know that eventually we won't retain any trace of the dead vat, but if we think we're using `ephemeral.vats.get(vatID).dead` to catch lingering references, then we'll need to create one of those at startup for each dead vat.

The test for this would probably be to extend your no-zombies test:

* have bootstrap retain a reference to the dynamic vat's root object
* add a second method on the bootstrap vat, which will send a message to the dynamic vat and `testLog` the result
* in the second pass, after the dynamic vat is dead, `queueToExport` the bootstrap vat to trigger the new method
* `await c.run()` and confirm that the result was a rejection

If `ephemeral.vats.get(vatID)` fails (because we didn't recreate it), I think we might get a kernel panic, which would be bad. In the later stages of this project, we wouldn't get that far, because we'd have either replaced the kernel object table entry with a tombstone, or pushed the obituary further upstream (so maybe vat-bootstrap wouldn't emit the `syscall.send` in the first place). But at this stage, vat-bootstrap is still perfectly able to send messages at the dead vat, and we need to make sure the kernel side remembers enough to deal with that properly.

Reply:
More or less this same concern came to me as I was waking up this morning.

Right now a vat will definitely panic the kernel by sending a message to the dead vat once all trace of it has been lost (this was the problem I was encountering last night), so we need to make sure that failure mode gets cleanly intercepted instead of letting the world come down around us.  I don't think a `{ dead: true }` record in `ephemeral.vats` is the right approach for several reasons: (a) once the persistent store has no record of the vat ever having existed, there's no way to create such a record on replay, (b) if we did keep a vestigial record of these (either in memory or in the db), it would still be a storage leak (albeit a small one) with the potential to eat all the memory, and (c) as you point out, a vat with a misbehaved (or no) liveslots layer can still manufacture references to non-existent objects in non-existent vats and produce a faulty syscall, and these still need to be handled safely.  I think the right solution is to reject bogus presence references coming out of a vat, regardless of how they got generated.

Reply:
hm, or perhaps we should just move to the step where we test for `!ephemeral.vats.has(vatID)` instead of `ephemeral.vats.get(vatID).dead`, and we remove it from `ephemeral.vats` instead of adding the `.dead` flag

Either way, the new test is still important.

Reply:
@warner I've added tests for sends to dead vats after replay (and fixed a bug surfaced by this), so please take another look.

Also, a question: we can have a send to a dead or non-existent vat as a result of dangling presences left behind in other vats that were created before the vat was terminated.  However, a send is a deliberate act.  Is it possible to get into a situation where a _notification_ to a dead vat can happen after replay?

Reply:
Yes, if the now-dead vat is holding a promise which can be resolved by the survivor. Simplest setup is:

```js
const p1 = makePromiseKit();
const { root, control } = await vatAdmin~.createVat(bundle);
root~.foo(p1.promise);
control~.terminate();
await control~.done();
p1.resolve();
```

What that really means is that there will be a kernel promise (for p1) with the late vat in the `subscribers` list. If we preemptively remove the dead vat from the subscribers of all the promises it knows about, then the notification will never be sent to them. So we can either leave everything in place and prepare to bounce messages/notifications if/when they arrive, or scan everything and remove it ahead of time. 

We've got a long-term target of leaving the kernel with no trace of the late vat. But "no trace" needs to include all c-lists. Once it's gone from the c-lists, no other vat will be able to reference it, and the threat of messages-to-the-dead will go away. But we can't do that yet because we don't have a way to retire object IDs. The kernel could remove the dead objects from the kernel's object-table, but it'd be pretty rude to unilaterally remove some other vat's c-list entry without giving them fair notice that using the old entry is a terminal offense.

It might be useful to replace the kernel object table (which maps object-id to owning vat) with an entry that points to `null` or something. Or a special vat which responds to all messages with a rejection (sort of like `/dev/null` or `/dev/full` or `/bin/false`), which might be tidy. Until then, the object table entries for dead objects will point to a vatid, and we'll need to remember that it's dead.

So the short-term target should be:
* the dead vat's `ephemeral.vats` entry is gone, manager is gone, almost all DB state is gone
* all Promises which were decided by the dead vat are rejected
* all run-queue deliveries aimed at the dead vat will bounce
* all run-queue notifications aimed at the dead vat will bounce
* all future deliveries/notifications that hit the dead vat will bounce

Reply:
The fix to bounce notifications is trivial (basically, the same thing I did with deliveries).  I will add a test for this case.

Even once the c-list entries have been expunged and the kernel object tables GC'd, there's still action that has to be taken to intercept these kinds of things.  Messages to the dead can still happen, because a vat could still send a message or notification to a vpid or presence id retained within the vat.  It's just that it will fail at syscall time (when it tries to translate vat slots to kernel slots) rather than at delivery time.

Reply:
Right, although at that point it's clearly bad behavior of the sending vat, equivalent to them randomly making up void/vpid values and using them in syscalls. Once the kernel tells the vat that a c-list entry is being removed, the vat should not use that void/vpid again, and doing so is vat-fatal.